src/probability/kernel/disintegration.lean

```diff
@@ -0,0 +1,407 @@
+  swap, { exact measurable_prod_mk_left ht, },
+  have h1 : {c : ℝ | (a, c) ∈ prod.map id f '' t} = f '' {c : Ω | (a, c) ∈ t},
+  { ext1 x,
+    simp only [prod_map, id.def, mem_image, prod.mk.inj_iff, prod.exists, mem_set_of_eq],
+    split,
+    { rintros ⟨a', b, h_mem, rfl, hf_eq⟩,
+      exact ⟨b, h_mem, hf_eq⟩, },
+    { rintros ⟨b, h_mem, hf_eq⟩,
+      exact ⟨a, b, h_mem, rfl, hf_eq⟩, }, },
+  have h2 : cond_kernel_real ρ' (c, a).snd = η' (c, a).snd,
+  { rw [kernel.piecewise_apply, if_pos ha], },
+  rw [h1, h2],
+end
+
+variables [nonempty Ω]
+
+/-- Conditional kernel of a measure on a product space: a Markov kernel such that
+`ρ : measure (α × Ω) = (ρ.fst : measure α) ⊗ₖ (cond_kernel ρ : kernel α Ω)` (up to augmentations
+of the spaces to make that expression valid: see `probability_theory.measure_eq_comp_prod`). -/
+noncomputable
+def cond_kernel (ρ : measure (α × Ω)) [is_finite_measure ρ] : kernel α Ω :=
+(exists_cond_kernel ρ unit).some
+
+instance (ρ : measure (α × Ω)) [is_finite_measure ρ] : is_markov_kernel (cond_kernel ρ) :=
+(exists_cond_kernel ρ unit).some_spec.some
+
+lemma kernel.const_unit_eq_comp_prod (ρ : measure (α × Ω)) [is_finite_measure ρ] :
+  kernel.const unit ρ
+    = (kernel.const unit ρ.fst) ⊗ₖ (kernel.prod_mk_left unit (cond_kernel ρ)) :=
+(exists_cond_kernel ρ unit).some_spec.some_spec
+
+/-- **Disintegration** of finite product measures on `α × Ω`, where `Ω` is Polish Borel. Such a
+measure can be written as the composition-product of the constant kernel with value `ρ.fst`
+(marginal measure over `α`) and a Markov kernel from `α` to `Ω`. We call that Markov kernel
+`probability_theory.cond_kernel ρ`. -/
+theorem measure_eq_comp_prod (ρ : measure (α × Ω)) [is_finite_measure ρ] :
+  ρ = ((kernel.const unit ρ.fst) ⊗ₖ (kernel.prod_mk_left unit (cond_kernel ρ))) () :=
+by rw [← kernel.const_unit_eq_comp_prod, kernel.const_apply]
+
+/-- **Disintegration** of constant kernels. A constant kernel on a product space `α × Ω`, where `Ω`
+is Polish Borel, can be written as the composition-product of the constant kernel with value `ρ.fst`
+(marginal measure over `α`) and a Markov kernel from `α` to `Ω`. We call that Markov kernel
+`probability_theory.cond_kernel ρ`. -/
+theorem kernel.const_eq_comp_prod (ρ : measure (α × Ω)) [is_finite_measure ρ]
+  (γ : Type*) [measurable_space γ] :
+  kernel.const γ ρ = (kernel.const γ ρ.fst) ⊗ₖ (kernel.prod_mk_left γ (cond_kernel ρ)) :=
+begin
+  ext a s hs : 2,
+  simpa only [kernel.const_apply, kernel.comp_prod_apply _ _ _ hs, kernel.prod_mk_left_apply']
+    using kernel.ext_iff'.mp (kernel.const_unit_eq_comp_prod ρ) () s hs,
+end
+
+lemma lintegral_cond_kernel (ρ : measure (α × Ω)) [is_finite_measure ρ]
+  {f : α × Ω → ℝ≥0∞} (hf : measurable f) :
+  ∫⁻ a, ∫⁻ ω, f (a, ω) ∂(cond_kernel ρ a) ∂ρ.fst = ∫⁻ x, f x ∂ρ :=
+begin
+  conv_rhs { rw measure_eq_comp_prod ρ, },
+  rw [kernel.lintegral_comp_prod _ _ _ hf, kernel.const_apply],
+  simp_rw kernel.prod_mk_left_apply,
+end
+
```

Comment:
It feels strange to have an extensive API for the real case but not for the general case. Are you planning to extend the API to the general case in another PR?

Reply:
I will definitely extend the API in future PRs. It is true that this PR gets to the definition and then pretty much stops there.